templates/commands/render_test.go

```diff
@@ -0,0 +1,164 @@
+// Copyright 2023 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package commands
+
+import (
+	"fmt"
+	"io/fs"
+	"testing"
+	"testing/fstest"
+
+	"github.com/abcxyz/pkg/testutil"
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+)
+
+func TestParseFlags(t *testing.T) {
+	t.Parallel()
+
+	cases := []struct {
+		name    string
+		args    []string
+		want    *Render
+		wantErr string
+	}{
+		{
+			name: "all_flags_present",
+			args: []string{
+				"--source", "helloworld@v1",
+				"--spec", "my_spec.yaml",
+				"--dest", "my_dir",
+				"--git-protocol", "https",
+				"--input", "x=y",
+				"--log-level", "info",
+				"--force-overwrite",
+				"--keep-temp-dirs",
+			},
+			want: &Render{
+				flagSource:         "helloworld@v1",
+				flagSpec:           "my_spec.yaml",
+				flagDest:           "my_dir",
+				flagGitProtocol:    "https",
+				flagInputs:         map[string]string{"x": "y"},
+				flagLogLevel:       "info",
+				flagForceOverwrite: true,
+				flagKeepTempDirs:   true,
+			},
+		},
+		{
+			name: "minimal_flags_present",
+			args: []string{
+				"-s", "helloworld@v1",
+			},
+			want: &Render{
+				flagSource:         "helloworld@v1",
+				flagSpec:           "./spec.yaml",
+				flagDest:           ".",
+				flagGitProtocol:    "https",
+				flagInputs:         nil,
+				flagLogLevel:       "warning",
+				flagForceOverwrite: false,
+				flagKeepTempDirs:   false,
+			},
+		},
+		{
+			name:    "required_flag_is_missing",
+			args:    []string{},
+			wantErr: "--source is required",
+		},
+	}
+
+	for _, tc := range cases {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+			r := &Render{}
+			err := r.parseFlags(tc.args)
+			if tc.wantErr != "" {
+				if diff := testutil.DiffErrString(err, tc.wantErr); diff != "" {
+					t.Fatal(diff)
+				}
+				return
+			}
```

Comment:
nit: I suppose it can still work like this?
```suggestion
			if diff := testutil.DiffErrString(err, tc.wantErr); diff != "" {
				t.Fatal(diff)
			}
```

Reply:
I tried that, but unfortunately it doesn't work. In the error case, some of the flags are parsed and some are not, so we need to skip the following call to cmp.Diff. We can't just assume that all flag fields will have a zero value after an error.